File: backend/src/main/java/br/ufpr/sept/so2/modules/presenca/domain/AttendanceMode.java
package br.ufpr.sept.so2.modules.presenca.domain;

import br.ufpr.sept.so2.shared.domain.exception.DadoInvalidoException;

public enum AttendanceMode {
    QR_SINGLE,
    QR_DUAL,
    SECRET_SINGLE,
    SECRET_DUAL;

    public boolean isDual() {
        return this == QR_DUAL || this == SECRET_DUAL;
    }

    public boolean isSecret() {
        return this == SECRET_SINGLE || this == SECRET_DUAL;
    }

    public boolean exercitadoNesteSprint() {
        return this == SECRET_SINGLE;
    }

    public static AttendanceMode from(String raw) {
        if (raw == null || raw.isBlank()) {
            throw new DadoInvalidoException("Modo de presença obrigatório.");
        }
        try {
            return AttendanceMode.valueOf(raw.trim().toUpperCase());
        } catch (IllegalArgumentException ex) {
            throw new DadoInvalidoException("Modo de presença inválido.");
        }
    }
}

File: backend/src/main/kotlin/br/ufpr/sept/so2/modules/presenca/domain/AttendanceMode.kt
package br.ufpr.sept.so2.modules.presenca.domain

import br.ufpr.sept.so2.shared.domain.exception.DadoInvalidoException

enum class AttendanceMode {
    QR_SINGLE,
    QR_DUAL,
    SECRET_SINGLE,
    SECRET_DUAL,
    ;

    fun isDual(): Boolean = this == QR_DUAL || this == SECRET_DUAL

    fun isSecret(): Boolean = this == SECRET_SINGLE || this == SECRET_DUAL

    fun exercitadoNesteSprint(): Boolean = this == SECRET_SINGLE

    companion object {
        fun from(raw: String?): AttendanceMode {
            if (raw.isNullOrBlank()) {
                throw DadoInvalidoException("Modo de presença obrigatório.")
            }
            return try {
                valueOf(raw.trim().uppercase())
            } catch (_: IllegalArgumentException) {
                throw DadoInvalidoException("Modo de presença inválido.")
            }
        }
    }
}
